clients/src/main/java/org/apache/kafka/common/utils/ByteBufferOutputStream.java

```diff
@@ -105,7 +105,7 @@ public int initialCapacity() {
         return initialCapacity;
     }
 
-    private void maybeExpandBuffer(int remainingRequired) {
+    public void maybeExpandBuffer(int remainingRequired) {
```

Comment:
Could we let `writeHeader.writeHeader` take the `ByteBufferOuputStream` instead of the `ByteBuffer` as the first parameter to avoid exposing this private function? It seems to be only used here and in the MemoryRecordsBuilder, whose caller has the `bufferStream` at hand as well.